fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/QueryToKeyMatcher.java

```diff
@@ -496,14 +508,20 @@ private KeyExpression extractKeyFromKeyWithValue(KeyWithValueExpression keyWithV
             // cover the entire query, so keep trying a more and more selective prefix of the key until
             // we no longer hit the unsplittable exception
             if (matchingMode == MatchingMode.SATISFY_QUERY) {
-                int adjustedSplitPoint = keyWithValue.getSplitPoint() - 1;
-                while (adjustedSplitPoint >= 0) {
-                    try {
-                        return keyWithValue.getInnerKey().getSubKey(0, adjustedSplitPoint);
-                    } catch (BaseKeyExpression.UnsplittableKeyExpressionException innerErr) {
-                        adjustedSplitPoint--;
-                    }
-                }
+                return extractPrefixUntilSplittable(keyWithValue.getInnerKey(), keyWithValue.getSplitPoint());
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private KeyExpression extractPrefixUntilSplittable(KeyExpression wholeKeyExpression, int originalSplitPoint) {
+        int adjustedSplitPoint = originalSplitPoint - 1;
+        while (adjustedSplitPoint >= 0) {
```

Comment:
Is it actually useful to have this succeed for `0` by returning `EmptyKeyExpression`?

Reply:
I think it never actually needs to return an `EmptyKeyExpression`, and that it could return `null` here. I added a test for the one case I could think where it might actually be needed (which is in the case where it can't pull any columns out of the `wholeKeyExpression`, but the predicate is also empty), and that also worked here (because the emptiness of the predicate is checked before the `QueryToKeyMatcher` is constructed).

I added the test, as that seems like a good regression test to have, but I also left this as is because it seems like it's easier to reason about this code if it returns an `EmptyKeyExpression` if it cannot extract any columns rather than having it return `null`. (Like, in the future, if we added a `Query.any()` component that evaluated to `true` on all inputs, then having this return `EMPTY` would be the correct thing to do here.)

Reply:
I can buy that, but should it then return `EMPTY` instead of `null` if the `try` fails? That is, do we want to distinguish the case where the key expression refuses to extras zero columns?